### DISCUSSION TITLE
Mojo::URL->child

### DISCUSSION THREAD

**Discussion by guest20** [mojolicious/mojo#2136] [General]
I buld a load of couchdb urls from request parameters:

```perl
my $couch = Mojo::URL->new(config->{couch_root} // 'http://localhost:5984');
get '/pet-shop/:animal/:id' => sub ($c) {
  $c->render_later;

  my $pets = $couch->clone; 
  push @{ $pets->path }, $c->stash('animal'), $c->stash('id');

  $c->ua->get_p( $pets )#
  ->then (sub ($tx) {
    $c->render( pet => $tx->res->json ) 
  });
};

```

A couple of my apps contain this:

```perl
sub Mojo::URL::child {
    my ($url,@bits) = @_;
    my $new = $url->clone;
    push @{ $new->path }, @bits;
    $new
}
```

Which means I can safely:

```perl
  my $pets = $couch->child($c->stash('animal'), $c->stash('id') );
  $c->ua->get_p( $pets )...
```

... which seems a bit nicer

Would a well tested patch like this be interesting to y'all?

**Comment by christopherraa** [comment]
I wonder if [Mojo::Path->parts](https://docs.mojolicious.org/Mojo/Path#parts) might be what you want. Perhaps this would work for you?

```perl
my $pets = $couch->clone;
$pets->path->parts([$c->param('animal'), $c->param('id')]);
$c->ua->get_p($pets)
  ->then(sub ($tx) { $c->render(pet => $tx->res->json) });
```

The above naturally wouldn't work if you already have path elements in the cloned URL. If that is the case those would have to be prepended.

If you are looking into this further and want something that would work with existing path elements then perhaps adding arrayref-support to `Mojo::Path->merge` makes sense. My only other suggestion would be to consider naming very carefully. Using `child` in a path-context sounds a bit off to me, though I am unable to say exactly _why_ I feel that way. Perhaps `$path->append([qw/foo bar baz/])` could be considered. At least to me "appending parts to a path" makes sense contextually.

I cannot speak to a patch being welcome or not, but if it has complete test coverage, follow the style requirements and don't have negative performance implications (this should be testet and methodology / code shared) then the chances are definitely much better than if those requirements are not met.

**Reply by guest20** [reply]
> wouldn't work if you already have path elements in the cloned URL.

It often does.

> Using child in a path-context sounds a bit off to me

It is in path context. It operates on `$url->path` in exactly the same way as `Mojo::File->child`.

> Perhaps $path->append

I haven't written tests yet, but I bet "appending" `..` isn't quite the right verbage. 
That said, `..` isn't a `child` either ...

**Reply by christopherraa** [reply]
You make valid points. I don't think I have much more to contribute to this discussion at this point though. I'd suggest you propose a change through a pull request, and then further discussion could be had there. Make sure to read through 
 the [contribution documentation](https://docs.mojolicious.org/Mojolicious/Guides/Contributing).
